Customize rest endpoints

vitorgamer58 opened the pull request:
<!-- Please ensure your PR title is brief and descriptive for a good changelog entry -->
<!-- Link to issue if there is one -->
<!-- markdownlint-disable -->

Fixes #28 and #47

<!-- markdownlint-restore -->

<!-- Describe what the changes are -->

## Proposed Changes

1. Custom endpoints for usecases (verb, path and resourceName)
2. Endpoints for `herbariu.crud.other` usecases

## Readiness Checklist

### Author/Contributor
- [x] If documentation is needed for this change, has that been included in this pull request
- [x] Remember to check if code coverage decrease, if so, please implement the necessary tests

### Reviewing Maintainer
- [ ] Label as `breaking` if this is a large fundamental change
- [ ] Label as either `automation`, `bug`, `documentation`, `enhancement`, `infrastructure`, or `performance`

vitorgamer58 committed:
feat: should customize REST endpoints

generateRoutes should receive metadata

https://github.com/herbsjs/herbs2rest/issues/28

vitorgamer58 committed:
feat: group custom endpoints and routes for crud.other

Group custom endpoints in a constant for clean code and endpoints for crud.other usecases

https://github.com/herbsjs/herbs2rest/issues/47

vitorgamer58 committed:
test(generateroutes): added tests and fixed broken tests

vitorgamer58 committed:
feat(generateroutes): others endpoints should have request

Endpoint verb of type 'other' must be displayed

vitorgamer58 committed:
test: added tests for custom endpoints

dalssoft commented on src/generateRoutes.js:
```diff
@@ -71,6 +79,21 @@ function generateRoutes(routes, app, endpointInfo = false) {
+
+    if (route.other) route.other.forEach((other) => {
+      if (other.REST?.verb && other.REST?.path) {
+        const endpoint = `${other.REST.path}`
```
There should be a default verb for use cases with `herbarium.crud.other` but without a explicit http verb. My suggestion is to be defaulted to `POST`

vitorgamer58 replied:
Fixed

dalssoft commented on src/generateControllers.js:
```diff
@@ -11,11 +11,14 @@ function generateControllers({ herbarium, controller = defaultController }) {
+            post: usecases.post ? { usecase: usecases.post.usecase, controller, REST: usecases.post.REST } : undefined,
+            put: usecases.put ? { usecase: usecases.put.usecase, controller, REST: usecases.put.REST } : undefined,
+            delete: usecases.del ? { usecase: usecases.del.usecase, controller, REST: usecases.del.REST } : undefined,
+            other: usecases.other.length ? usecases.other.map((usecase) => {
```
It seems `controllerInfos` needs a `group` in order to work. What is a use case has no `group`? Ex: `GET /search - Search all entities in your domain`

vitorgamer58 replied:
A usecase without a group will not have an endpoint, in this case.

dalssoft replied:
We need to think in a structure that deals with:
- [x] CRUD use cases - Clear action and clear resource.
- [ ] Extra use cases for a resource - need extra metadata (Ex: `herbarium.usecases.get('UpdateUserPermission').metadata({ REST: { path: '/user/:id/permission' } })`)
- [ ] Use cases not linked to a resource (ex: `/search`)

Here the resource = group, since the current convention uses the group as the `resourceName`.

vitorgamer58 replied:
I'll take a look in this cases.

vitorgamer58 replied:
Ok, now the code works if an usecase is created without linked to a resource (group and resourceName), but REST.path should be passed.

vitorgamer58 commented on src/generateRoutes.js:
```diff
@@ -71,6 +79,21 @@ function generateRoutes(routes, app, endpointInfo = false) {
     }
+
+    if (route.other) route.other.forEach((other) => {
+      if (other.REST?.verb && other.REST?.path) {
```
Does this refer to how this is exported in metadata?
https://github.com/herbsjs/herbs2rest/blob/aa77d220cc18c34389a9b592402723d384c49658/test/generateControllers.test.js#L99

As for the if, I programmed it so that only one endpoint is generated for the usecase, if it is exported with path and verb metadata, I'm still working on the so resourceName option.

If a usecase has no path or verb, and is of type other, it is likely that the user would not want an endpoint created.

Soon I will work on the standard POST verb for usecases of type other.

vitorgamer58 replied:
Perhaps we can wrap the herbarium in this, and pass the verb as` herbarium.verb.{getAll | getById | post | put | delete}` in the usecase export, or `herbarium.other`, but this when exporting the usecase, since in this line we only read the metadata of the found usecases.

dalssoft replied:
My point is just that the current UX is `{action} = {path}`, where `{action}` being `getAll | getById | post | put | delete`. The `{action}` will be later translated to a HTTP Verb (`GET`, `POST`, etc)

vitorgamer58 replied:
Ok, so instead of `verb: "GET"` we can pass something like:
`herbarium.actions.{getAll | getById | post | put | delete}` ?

dalssoft replied:
As mentioned in https://github.com/herbsjs/herbs2rest/pull/49#discussion_r1090069003, it is just 
`other.REST.{getAll | getById | post | put | delete} = '/user/:id/bla/foo'`, no `.actions.` in it.

vitorgamer58 replied:
Solved

vitorgamer58 committed:
feat(generateroutes): added support for custom resourceName

Added tests for custom routes (verb, resourceName and path) and minor fixes

vitorgamer58 committed:
feat(generateroutes): added default verb for usecases other

When the http verb is not passed in the usecase metadata, a default POST verb should be used.

vitorgamer58 committed:
feat(generateroutes): endpoint logic change and minor changes

When path is passed, must ignore resourceName
 Endpoint for other usecases should be created if
path or resourceName is passed
 Improved unit tests

vitorgamer58 committed:
feat(generateroutes): id field option should be passed by usecase in case of custom endpoint

dalssoft commented on test/generateControllers.test.js:
```diff
@@ -146,3 +150,243 @@ describe('Herbs2Rest - Generate Routes With Herbarium', () => {
+
+  beforeEach(() => usecaseTest())
+
+  it('Should resolve and create a custom get by id route after add inside Herbarium usecase list', (done) => {
```
this is a test for a path override, right?

suggestion:
```js
describe('Generate Custom Routes With Herbarium', ()
   it('get - should ignore the conventions and use the given path'
```

vitorgamer58 replied:
I Agree

vitorgamer58 commented on src/generateRoutes.js:
```diff
@@ -8,12 +8,28 @@ function generateRoutes(routes, app, endpointInfo = false) {
       [idFieldName] = Object.entries(route.entity.prototype.meta.schema).find(([_key, value]) => value?.options.isId) || []
     }
 
+    const customEndpoints = {
```
I Agree

vitorgamer58 commented on src/generateRoutes.js:
```diff
@@ -8,12 +8,28 @@ function generateRoutes(routes, app, endpointInfo = false) {
+      delete: route.delete?.REST?.path,
+    }
+
+    const resourceName = {
```
I Agree

vitorgamer58 commented on src/generateRoutes.js:
```diff
@@ -8,12 +8,28 @@ function generateRoutes(routes, app, endpointInfo = false) {
+
     if (route.getAll) {
-      const endpoint = `/${route.name}`
+      const endpoint = customEndpoints.getAll || `/${resourceName.getAll}`
```
I Agree

vitorgamer58 committed:
refactor(generateroutes): changed variable names and minor fixes

Changed customEndpoints to forcedPath
Changed resourceName to generatedPath
Changed endpoint to path
Changed responsibility for the generated path from the path variable to generatedPath

vitorgamer58 commented on src/generateRoutes.js:
```diff
@@ -8,12 +8,28 @@ function generateRoutes(routes, app, endpointInfo = false) {
 
     let idFieldName = null
-    if(route.entity){
+    if (route.entity) {
```
I believe that if I wrote tests in generateRoutes, they would be the same test cases, with the same metadata, wouldn't it be testing the same thing twice in two different places?
The tests in generateControllers generate routes with generateRoutes, that's why, even though I made big changes in generateRoutes, the code coverage didn't decrease, because the tests in generateControllers are covering 100% of the cases of generateRoutes.
If you think it makes sense, I can pass the cases from generateControllers to generateRoutes, and then make a slightly different assertion in generateControllers.

vitorgamer58 committed:
refactor(generatecontrollers.test): changed name os tests and changed beforeEach to before

vitorgamer58 committed:
style(generateroutes.test): removed comma at the end of test cases and prettier

vitorgamer58 committed:
feat(generateroutes.js): verb and Path change

The verb must be the key and the value must be the path, in the usecase metadata REST object

vitorgamer58 committed:
docs(readme): add custom endpoints documentation to README

vitorgamer58 committed:
test(generatecontrollers): fix incorrect test

vitorgamer58 committed:
docs(readme): added examples and some refactor

vma-vortx commented on src/generateRoutes.js:
```diff
@@ -61,17 +75,112 @@ function generateRoutes(routes, app, endpointInfo = false) {
+    delete: getVerb(route.delete?.REST)
+  }
+
+  if (verbs.getAll) {
```
Those "if's" could be refactored since all do the same just changing the verb used.
